packages/loader/container-loader/src/connectionManager.ts

```diff
@@ -647,18 +647,21 @@ export class ConnectionManager implements IConnectionManager {
 
 				const waitStartTime = performance.now();
 				const retryDelayFromError = getRetryDelayFromError(origError);
-				if (retryDelayFromError !== undefined) {
-					// If the error told us to wait, then we wait.
-					this.props.reconnectionDelayHandler(retryDelayFromError, origError);
+				// If the error told us to wait, then we wait for that time. If the error didn't tell us to wait,
+				// let's still wait a little bit before retrying. We skip this delay if we're confident we're offline,
+				// because we probably just need to wait to come back online. But we never have strong signal of being
+				// offline, so we at least wait for sometime.
+				if (retryDelayFromError !== undefined || globalThis.navigator?.onLine !== false) {
+					delayMs = Math.max(retryDelayFromError ?? 0, delayMs);
+					// Raise event in case the delay was there.
+					this.props.reconnectionDelayHandler(delayMs, origError);
 					await new Promise<void>((resolve) => {
-						setTimeout(resolve, retryDelayFromError);
+						setTimeout(resolve, delayMs);
 					});
-				} else if (globalThis.navigator?.onLine !== false) {
-					// If the error didn't tell us to wait, let's still wait a little bit before retrying.
-					// We skip this delay if we're confident we're offline, because we probably just need to wait to come back online.
+					delayMs = Math.min(delayMs * 2, calculateMaxWaitTime(origError));
```

Comment:
Suggesting: Every place that uses calculateMaxWaitTime does the same thing - doubles previous value and uses min().  I'd suggest moving this logic into calculateMaxWaitTime, such that this code would look like

delayMs = calculateMaxWaitTime(delayMs , origError);

This would also make it easier if in the future we want to change logic. And less likely that someone will make mistake and would not use min()

Reply:
maybe there is a chance to do even more code sharing - I think all code paths also do the equivalent of

delayMs = Math.max(retryDelayFromError ?? 0, delayMs);

Maybe we can hide all this logic in one function? Not sure (as we need both current delay and future delay).
If we divide initial values by half (i.e. MaxReconnectDelayInMsWhenEndpointIsReachable), then I think we can have just one line at the top (before wait):

```ts
delayMs = calculateMaxWaitTime(delayMs, origError);
```

Where calculateMaxWaitTime() would do this:

```ts
export function calculateMaxWaitTime(delayMs, error: unknown): number {
        let retryDelayFromError = getRetryDelayFromError(origError);
        delayMs = Math.max(retryDelayFromError ?? 0, delayMs * 2);
	delayMs = Math.min(
               delayMs,
               isFluidError(error) && error.getTelemetryProperties().endpointReached === true
		   ? MaxReconnectDelayInMsWhenEndpointIsReachable
		   : MaxReconnectDelayInMsWhenEndpointIsNotReachable;
}
```

What do you think?

Reply:
But that's all about readability and maintanability of the code, your current changes (patches) are good to go I believe